code/modules/research/machinery/_production.dm

```diff
@@ -73,17 +83,109 @@
 		reagents.trans_to(G, G.reagents.maximum_volume)
 	return ..()
 
-/obj/machinery/rnd/production/proc/do_print(path, amount, list/matlist, notify_admins)
+/obj/machinery/rnd/production/proc/add_to_queue(datum/design/D, amount, notify_admins)
+	if(length(queue) >= MAX_QUEUE_LEN)
+		say("The build queue is full.")
+		playsound(src, 'sound/machines/buzz-two.ogg', 50, FALSE)
+		return
+
+	queue += list(list(D, amount))
+
 	if(notify_admins)
-		investigate_log("[key_name(usr)] built [amount] of [path] at [src]([type]).", INVESTIGATE_RESEARCH)
-		message_admins("[ADMIN_LOOKUPFLW(usr)] has built [amount] of [path] at \a [src]([type]).")
+		investigate_log("[key_name(usr)] queued [amount] of [D.build_path] at [src]([type]).", INVESTIGATE_RESEARCH)
+		message_admins("[ADMIN_LOOKUPFLW(usr)] has queued [amount] of [D.build_path] at \a [src]([type]).")
+
+	if(!busy)
+		run_queue()
+
+/obj/machinery/rnd/production/proc/run_queue()
+	set waitfor = FALSE
+	if(busy)
+		return
+	busy = TRUE
+
+	while(length(queue))
+		var/list/queue_packet = queue[1]
+		queue -= list(queue_packet)
+		processing_packet = queue_packet
+		updateUsrDialog()
+		do_print(queue_packet[1], queue_packet[2])
+
+	processing_packet = null
+	busy = FALSE
+	updateUsrDialog()
+
+/obj/machinery/rnd/production/proc/do_print(datum/design/D, amount)
+	if(!can_build_design(D, amount))
+		return FALSE
+
+	var/atom/path = D.build_path
+
+	var/coeff = efficient_with(D.build_path) ? efficiency_coeff : 1
+	var/list/efficient_mats = list()
+	for(var/MAT in D.materials)
+		efficient_mats[MAT] = D.materials[MAT]/coeff
+
+	var/power = active_power_usage
+	amount = clamp(amount, 1, 50)
+
+	for(var/M in D.materials)
+		power += round(D.materials[M] * amount / 35)
+
+	power = min(active_power_usage, power)
+	use_power(power)
+
+	materials.mat_container.use_materials(efficient_mats, amount)
+	materials.silo_log(src, "built", -amount, "[D.name]", efficient_mats)
+
+	for(var/R in D.reagents_list)
+		reagents.remove_reagent(R, D.reagents_list[R]*amount/coeff)
+
+	var/time = (((D.construction_time || 2 SECONDS) / efficiency_coeff) * amount) ** 0.8
+	if(!do_after(src, src, time, DO_PUBLIC|IGNORE_USER_LOC_CHANGE, display = image(initial(path.icon), initial(path.icon_state))))
```

Comment:
...I want pictures of this because I'm **REALLY** not sure how I feel about this.